File: modules/dms-vision/ios/DmsRecords.swift
// The argument records of `start` and `setPolicy`. Their field names are EXACTLY the keys of
// src/wire.ts `startOptionsSchema` / `capturePolicySchema` (native-ios.test.ts compares them).
// Every field is optional here so a missing or mistyped value is refused by `validated()` with
// E_BAD_ARGS, never by the bridge with a generic error.

import CryptoKit
import ExpoModulesCore
import Foundation

struct StartOptionsRecord: Record {
  @Field var gateToken: String?
  @Field var fps: Double?
  @Field var gazeNet: Bool?
  @Field var gazeNetEvery: Double?
  @Field var delegate: String?
  @Field var rotationOffsetDegrees: Double?

  func validated() throws -> (token: String, fps: Int, gazeNet: Bool, every: Int, gpu: Bool, rotationOffset: Int) {
    guard let token = gateToken, !token.isEmpty else { throw DmsError.badArgs("a gate token is required") }
    guard let f = fps, DmsConstants.ALLOWED_FPS.map { Double($0) }.contains(f) else { throw DmsError.badArgs("fps") }
    guard let net = gazeNet else { throw DmsError.badArgs("gazeNet") }
    guard let e = gazeNetEvery, e == 1 || e == 2 else { throw DmsError.badArgs("gazeNetEvery") }
    guard let d = delegate, d == "cpu" || d == "gpu" else { throw DmsError.badArgs("delegate") }
    guard let r = rotationOffsetDegrees, DmsConstants.ALLOWED_ROTATIONS.map { Double($0) }.contains(r) else {
      throw DmsError.badArgs("rotationOffsetDegrees")
    }
    return (token, Int(f), net, Int(e), d == "gpu", Int(r))
  }
}

struct CapturePolicyRecord: Record {
  @Field var gateToken: String?
  @Field var capture: String?
  @Field var fps: Double?
  @Field var gazeNet: Bool?
  @Field var gazeNetEvery: Double?
  @Field var setupMode: Bool?
  @Field var previewAllowed: Bool?

  func validated() throws -> (token: String, capture: String, fps: Int, gazeNet: Bool, every: Int, setupMode: Bool, previewAllowed: Bool) {
    guard let token = gateToken, !token.isEmpty else { throw DmsError.badArgs("a gate token is required") }
    guard let c = capture, c == "run" || c == "pause" else { throw DmsError.badArgs("capture") }
    guard let f = fps, DmsConstants.ALLOWED_FPS.map { Double($0) }.contains(f) else { throw DmsError.badArgs("fps") }
    guard let net = gazeNet else { throw DmsError.badArgs("gazeNet") }
    guard let e = gazeNetEvery, e == 1 || e == 2 else { throw DmsError.badArgs("gazeNetEvery") }
    guard let setup = setupMode, let preview = previewAllowed else { throw DmsError.badArgs("setupMode / previewAllowed") }
    return (token, c, Int(f), net, Int(e), setup, preview)
  }
}

extension DmsVisionModule {
  /// sha256 of a bundled file (diagnostics only; computed on request, never per frame).
  static func sha256(path: String) -> String? {
    guard let data = try? Data(contentsOf: URL(fileURLWithPath: path)) else { return nil }
    return SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
  }
}

File: modules/dms-vision/android/src/main/java/expo/modules/dmsvision/DmsRecords.kt
// The argument records of `start` and `setPolicy`. Their field names are EXACTLY the keys of
// src/wire.ts `startOptionsSchema` / `capturePolicySchema` (native-android.test.ts compares them).
// Every field is nullable here so a missing or mistyped value is refused by `validated()` with
// E_BAD_ARGS, never by the bridge with a generic error.

package expo.modules.dmsvision

import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class StartOptions(val token: String, val fps: Int, val gazeNet: Boolean, val every: Int, val gpu: Boolean, val rotationOffset: Int)

class StartOptionsRecord : Record {
  @Field var gateToken: String? = null
  @Field var fps: Double? = null
  @Field var gazeNet: Boolean? = null
  @Field var gazeNetEvery: Double? = null
  @Field var delegate: String? = null
  @Field var rotationOffsetDegrees: Double? = null

  fun validated(): StartOptions {
    val token = gateToken?.takeIf { it.isNotEmpty() } ?: throw DmsError.badArgs("a gate token is required")
    val f = fps?.takeIf { v -> DmsConstants.ALLOWED_FPS.any { it.toDouble() == v } } ?: throw DmsError.badArgs("fps")
    val net = gazeNet ?: throw DmsError.badArgs("gazeNet")
    val e = gazeNetEvery?.takeIf { it == 1.0 || it == 2.0 } ?: throw DmsError.badArgs("gazeNetEvery")
    val d = delegate?.takeIf { it == "cpu" || it == "gpu" } ?: throw DmsError.badArgs("delegate")
    val r = rotationOffsetDegrees?.takeIf { v -> DmsConstants.ALLOWED_ROTATIONS.any { it.toDouble() == v } }
      ?: throw DmsError.badArgs("rotationOffsetDegrees")
    return StartOptions(token, f.toInt(), net, e.toInt(), d == "gpu", r.toInt())
  }
}

class CapturePolicy(
  val token: String, val capture: String, val fps: Int, val gazeNet: Boolean, val every: Int,
  val setupMode: Boolean, val previewAllowed: Boolean
)

class CapturePolicyRecord : Record {
  @Field var gateToken: String? = null
  @Field var capture: String? = null
  @Field var fps: Double? = null
  @Field var gazeNet: Boolean? = null
  @Field var gazeNetEvery: Double? = null
  @Field var setupMode: Boolean? = null
  @Field var previewAllowed: Boolean? = null

  fun validated(): CapturePolicy {
    val token = gateToken?.takeIf { it.isNotEmpty() } ?: throw DmsError.badArgs("a gate token is required")
    val c = capture?.takeIf { it == "run" || it == "pause" } ?: throw DmsError.badArgs("capture")
    val f = fps?.takeIf { v -> DmsConstants.ALLOWED_FPS.any { it.toDouble() == v } } ?: throw DmsError.badArgs("fps")
    val net = gazeNet ?: throw DmsError.badArgs("gazeNet")
    val e = gazeNetEvery?.takeIf { it == 1.0 || it == 2.0 } ?: throw DmsError.badArgs("gazeNetEvery")
    val setup = setupMode ?: throw DmsError.badArgs("setupMode")
    val preview = previewAllowed ?: throw DmsError.badArgs("previewAllowed")
    return CapturePolicy(token, c, f.toInt(), net, e.toInt(), setup, preview)
  }
}
